src/lib.rs

```diff
@@ -37,18 +39,46 @@ macro_rules! assert_matches {
                 e, stringify!($pat), stringify!($cond))
         }
     };
-    ( $e:expr , $pat:pat , $($arg:tt)* ) => {
+    ( $e:expr , $pat:pat => $($arm:expr),* ) => {
+        match $e {
+            $pat => ($($arm),*),
+            ref e => panic!("assertion failed: `{:?}` does not match `{}`",
+                e, stringify!($pat))
+        }
+    };
+    ( $e:expr , $pat:pat if $cond:expr => $($arm:expr),* ) => {
+        match $e {
+            $pat if $cond => ($($arm),*),
+            ref e => panic!("assertion failed: `{:?}` does not match `{} if {}`",
+                e, stringify!($pat), stringify!($cond))
+        }
+    };
+    ( $e:expr , $pat:pat , $($arg:expr),* ) => {
         match $e {
             $pat => (),
             ref e => panic!("assertion failed: `{:?}` does not match `{}`: {}",
-                e, stringify!($pat), format_args!($($arg)*))
+                e, stringify!($pat), format_args!($($arg),*))
         }
     };
-    ( $e:expr , $pat:pat if $cond:expr , $($arg:tt)* ) => {
+    ( $e:expr , $pat:pat if $cond:expr , $($arg:expr),* ) => {
```

Comment:
Why are you changing this? This is not backward compatible and it would prevent users from using the following valid construct: `assert_matches!(foo, _, "{value:?}", value=foo);`

Reply:
To avoid `error: local ambiguity: multiple parsing options: built-in NTs tt ('arg') or expr ('arm').`. I did not think about this use case... I hoped this would be backwards compatible but looks like it can't be. I will experiment more with this - perhaps there is a way.